File: 백준/Silver/1748. 수 이어 쓰기 1/수 이어 쓰기 1.cc

#include <iostream>
#include <vector> 
#include <string> 
#include <cstring> 
#include <algorithm> 
#include <cmath> 
#include <unordered_map>
#include <map> 
#include <queue> 
#include <set> 
#define jl "\n"
using namespace std; 

int N ; 

long long pref[12]  ;

int main() { 

    cin >> N ; 
    memset(pref, 0, sizeof(pref)) ;  

    pref[1] = 9 ;
    long long x = pref[1]; 
    for(int i=2 ; i <=9 ; i++){
        x = x*10 ; 
        pref[i] = x ; 
    } 
    

    int Nlen = to_string(N).length() ; 
    long long ans = 0 ; 
    for(int i=1 ; i <Nlen ; i++ ){
        ans +=pref[i]*i ;
    }
    

    if(N != pow(10,Nlen-1)){
        ans += (N- pow(10,Nlen-1)+1)*(Nlen) ;
    }else{
        ans +=(Nlen) ;
    }

    cout << ans << endl ; 

}

File: 백준/Silver/1748. 수 이어 쓰기 1/수 이어 쓰기 1.java
import java.util.Scanner ; 



public class Main  {
    public static void main(String[] args) {
        
        Scanner sc = new Scanner(System.in) ; 

        
        int N = sc.nextInt() ; 

        int nlen = Integer.toString(N).length() ;

        
        int ans =0 ; 
        int x = 9 ;
        for(int i=1  ; i < nlen ; i++){
            if(i >1) x *=10 ;
            ans += x*i ; 
        }

        if(N != Math.pow(10, nlen-1)){
            ans += (N- Math.pow(10, nlen-1)+1 )*nlen ; 
        } else {
            ans += nlen ;
        }
        System.out.println(ans) ; 
    }
}
